src/webui/api/torrentscontroller.cpp

```diff
@@ -252,6 +254,25 @@ namespace
             idList << BitTorrent::TorrentID::fromString(hash);
         return idList;
     }
+
+    nonstd::expected<QUrl, QString> validateWebseedUrl(const QString &urlStr)
+    {
+        const QString normalizedUrlStr = QUrl::fromPercentEncoding(urlStr.toUtf8());
+
+        // sanity check before parsing to prevent QUrl from checking for the existence of files on the filesystem
```

Comment:
>prevent QUrl from checking for the existence of files on the filesystem

What makes you think `QUrl` do something other than parsing/constructing URLs? Is this declared somewhere in the documentation? Or did you discover it by experience?

Reply:
This dates from when I was originally using `QUrl::fromUserInput`, which states:

> In the case the string corresponds to a valid file path on the system, a file:// URL is constructed, using [QUrl::fromLocalFile](https://doc.qt.io/qt-6/qurl.html#fromLocalFile)().

It's not clear to me whether a "valid file path" is determined based solely on e.g. character set or if the file system is checked. It's also not clear to me if the `QUrl` constructor has this same behavior.

Reply:
It looks like `QUrl::fromLocalFile()` [does indeed](https://github.com/qt/qtbase/blob/3642ce6e4b1912a774cc55c9a6a85ecf716d7080/src/corelib/io/qurl.cpp#L3822-L3830) check for the existence of a local file when a file URI is specified. However, the `QUrl` [parser](https://github.com/qt/qtbase/blob/3642ce6e4b1912a774cc55c9a6a85ecf716d7080/src/corelib/io/qurl.cpp#L1367) doesn't appear to exhibit this same behavior, instead relying solely on parsing the string's characters. So we're safe to perform the scheme check only once after passing the string to the `QUrl` constructor. There is of course no guarantee that the behavior of the constructor will remain the same in the future.

Reply:
> It looks like `QUrl::fromLocalFile()` [does indeed](https://github.com/qt/qtbase/blob/3642ce6e4b1912a774cc55c9a6a85ecf716d7080/src/corelib/io/qurl.cpp#L3822-L3830) check for the existence of a local file when a file URI is specified.

You're confused here. `QUrl::fromUserInput()` does actually do this (take a closer look at the code snippet you referred above), and this seems to be the only (exceptional) case where `QUrl` (not `QUrl` essentially, see below) does this (this behavior is explicitly mentioned in the documentation).
The `QUrl` class still remains just a parser/builder. `QUrl::fromUserInput()` is in fact not an integral part of it, but only an auxiliary "smart" function placed in its namespace.

Reply:
Sorry, I meant to say `fromUserInput` rather than `fromLocalFile`.

Reply:
> So we're safe to perform the scheme check only once after passing the string to the QUrl constructor.

I've updated the code to reflect this.